### PR TITLE
Add Infer plugin

### PR DESCRIPTION
**Pull request by geralt-encore**
https://github.com/pushtorefresh/storio/issues/381

I've added [this](https://github.com/uber-common/infer-plugin) Gradle plugin, which plays with Infer nicely.
It runs ok on my local machine. But we still need to install Infer to the CI server (not sure how) and then add two Gradle tasks to the CI script: `inferDebug` and `eradicateDebug` (for [Eradicate](http://fbinfer.com/docs/eradicate.html) feature).
Right now Infer doesn't see any issues (for testing purposes I've added dummy method with a silly null error and it has found it - so it actually works). Eradicate is finding one issue in storio-common module and fails right after that, so I can't say anything about other modules (can just run eradicate on separate modules, but haven't done it yet). If you are interested in applying these checks I could fix them before the merge.

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
But you don't run it during CI build. See https://github.com/artem-zinnatullin/qualitymatters/issues/21 :(

**Comment by geralt-encore** [comment]
That sucks =(
I've found only [this](https://github.com/facebook/infer/issues/153) solution, but it looks really bad.
Maybe it is an option to you it just manually? Better than nothing at least and could be helpful sometimes.

**Comment by artem-zinnatullin** [comment]
Okay, I'm fine with running it locally

**Comment by geralt-encore** [comment]
All eradicate warnings are fixed now

**Comment by nikitin-da** [comment]
Nice! 
👍 but covarage

**Comment by geralt-encore** [comment]
Yeah, but there is not much that I can do since I've added only annotation and couple of null checks

**Comment by nikitin-da** [comment]
Hm, I think false branches of that checks weren't covered =) https://codecov.io/gh/pushtorefresh/storio/compare/29a3b1348fd8e0b8d8b49227dd4288454e509099...7529dfe424c59a30bb21de67886bc05f9e484d92

**Comment by geralt-encore** [comment]
But most of these ifs are redundant actually and mostly about making infer stop complaining about them. So there is not much sense for covering them.

File: build.gradle [hunk 1]
```diff
@@ -17,6 +17,8 @@ buildscript {
         // https://github.com/palantir/gradle-jacoco-coverage
         classpath 'com.palantir:jacoco-coverage:0.2.0'
 
+        classpath "com.uber:infer-plugin:0.7.1"
```
Comment (artem-zinnatullin) [inline build.gradle]: single quotes pls

File: storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/delete/PreparedDeleteCollectionOfObjects.java [hunk 2]
```diff
@@ -96,7 +96,7 @@
                     final DeleteResult deleteResult = explicitDeleteResolver.performDelete(storIOContentResolver, object);
                     results.put(object, deleteResult);
                 }
-            } else {
+            } else if (objectsAndDeleteResolvers != null) {
```
Comment (artem-zinnatullin) [inline storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/delete/PreparedDeleteCollectionOfObjects.java]: no, it's ok, we better crash if it'll be null

Comment (artem-zinnatullin) [inline storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/delete/PreparedDeleteCollectionOfObjects.java]: just suppress it here

File: storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/put/PutResults.java [hunk 3]
```diff
@@ -65,7 +65,7 @@ public int numberOfInserts() {
         int numberOfInserts = 0;
 
         for (T object : results.keySet()) {
-            if (results.get(object).wasInserted()) {
+            if (results.get(object) != null && results.get(object).wasInserted()) {
```
Comment (artem-zinnatullin) [inline storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/put/PutResults.java]: result can never be null, we don't need this `!= null` here

File: storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/put/PutResults.java [hunk 4]
```diff
@@ -92,7 +92,7 @@ public int numberOfUpdates() {
         for (T object : results.keySet()) {
             final PutResult putResult = results.get(object);
 
-            if (putResult.wasUpdated()) {
+            if (putResult != null && putResult.wasUpdated() && putResult.numberOfRowsUpdated() != null) {
```
Comment (artem-zinnatullin) [inline storio-content-resolver/src/main/java/com/pushtorefresh/storio/contentresolver/operations/put/PutResults.java]: result can never be null, we don't need this `!= null` here

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteCollectionOfObjects.java [hunk 5]
```diff
@@ -115,7 +115,7 @@
                             lowLevel.notifyAboutChanges(Changes.newInstance(deleteResult.affectedTables()));
                         }
                     }
-                } else {
+                } else if (objectsAndDeleteResolvers != null){
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteCollectionOfObjects.java]: same as above

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PreparedPutCollectionOfObjects.java [hunk 6]
```diff
@@ -108,7 +108,7 @@
                             lowLevel.notifyAboutChanges(Changes.newInstance(putResult.affectedTables()));
                         }
                     }
-                } else {
+                } else if (objectsAndPutResolvers != null) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PreparedPutCollectionOfObjects.java]: same as above

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResults.java [hunk 7]
```diff
@@ -65,7 +65,7 @@ public int numberOfInserts() {
         int numberOfInserts = 0;
 
         for (T object : results.keySet()) {
-            if (results.get(object).wasInserted()) {
+            if (results.get(object) != null && results.get(object).wasInserted()) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResults.java]: result can never be null, we don't need this `!= null` here

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResults.java [hunk 8]
```diff
@@ -92,7 +92,7 @@ public int numberOfUpdates() {
         for (T object : results.keySet()) {
             final PutResult putResult = results.get(object);
 
-            if (putResult.wasUpdated()) {
+            if (putResult != null && putResult.wasUpdated() && putResult.numberOfRowsUpdated() != null) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResults.java]: result can never be null, we don't need this `!= null` here

**Comment by geralt-encore** [comment]
@artem-zinnatullin the thing is that it is impossible to suppress warnings on infer side, so without these redundant checks build is failing. Looks like infer is not smart enough for some cases.

**Comment by artem-zinnatullin** [comment]
@geralt-encore I definitely don't like that it changes logic in some places and instead of failing with NPE we're changing behavior.

What if you add plain old java `assert value != null`? It won't be executed on Android but will be executed in our JVM tests and probably will prevent Infer from complaining

**Comment by geralt-encore** [comment]
Yeah, I completely agree with you. I'll try assert solution and if it won't work we will probably have to abandon this idea =(

**Comment by geralt-encore** [comment]
Looks like assert stuff worked. I'll rework it then in a short time.

**Comment by artem-zinnatullin** [comment]
ok, great!

File: storio-common/build.gradle [hunk 9]
```diff
@@ -1,5 +1,6 @@
 apply plugin: 'com.android.library'
 apply plugin: 'checkstyle'
+apply plugin: 'com.uber.infer.android'
```
Comment (artem-zinnatullin) [inline storio-common/build.gradle]: Maybe apply it in root `build.gradle` for all `android` subprojects?

See https://github.com/pushtorefresh/storio/blob/592368c97524f736e315d7e89cf4f2b339c657b7/build.gradle#L107

File: storio-content-resolver-annotations-processor/build.gradle [hunk 10]
```diff
@@ -1,4 +1,5 @@
 apply plugin: 'java'
+apply plugin: 'com.uber.infer.java'
```
Comment (nikitin-da) [inline storio-content-resolver-annotations-processor/build.gradle]: What about common and sqlite annotation processor modules? ;)

Comment (geralt-encore) [inline storio-content-resolver-annotations-processor/build.gradle]: Thank you, I just forgot to remove it from here. It doesn't work with annotation processors at all.
```
* What went wrong:
Execution failed for task ':storio-content-resolver-annotations-processor:inferPrepare'.
> Configuration with name 'provided' not found.
```

**Comment by geralt-encore** [comment]
I have no idea what is the problem with `PutResultsTest` `checkToStringImplementation`. Something relate to assertions I guess.

**Comment by artem-zinnatullin** [comment]
lol wtf

**Comment by artem-zinnatullin** [comment]
Got it, your pr is #666

**Comment by geralt-encore** [comment]
Damn knew it would be a problem 😬

**Comment by artem-zinnatullin** [comment]
Still fails :(

**Comment by geralt-encore** [comment]
I have no idea how can I possibly fix it

**Comment by geralt-encore** [comment]
@artem-zinnatullin probably we can close it since no one has an idea what's wrong with failing test and it's not really important one.

**Comment by artem-zinnatullin** [comment]
Yeah, ok :(

**Pull request closed without merging**